web/api/tests/valid_urls.c

```diff
@@ -0,0 +1,777 @@
+    va_list args;
+    va_start(args, fmt);
+    printf("  INFO: ");
+    printf(fmt, args);
+    printf("\n");
+    va_end(args);
+}
+
+void __wrap_error_int(
+    const char *prefix, const char *file, const char *function, const unsigned long line, const char *fmt, ...)
+{
+    (void)prefix;
+    (void)file;
+    (void)function;
+    (void)line;
+    va_list args;
+    va_start(args, fmt);
+    printf("  ERROR: ");
+    printf(fmt, args);
+    printf("\n");
+    va_end(args);
+}
+
+void __wrap_fatal_int(const char *file, const char *function, const unsigned long line, const char *fmt, ...)
+{
+    (void)file;
+    (void)function;
+    (void)line;
+    va_list args;
+    va_start(args, fmt);
+    printf("FATAL: ");
+    printf(fmt, args);
+    printf("\n");
+    va_end(args);
+    fail();
+}
+
+WEB_SERVER_MODE web_server_mode = WEB_SERVER_MODE_STATIC_THREADED;
+char *netdata_configured_web_dir = "UNKNOWN FIXME";
+RRDHOST *localhost = NULL;
+
+struct config netdata_config = { .sections = NULL,
+                                 .mutex = NETDATA_MUTEX_INITIALIZER,
+                                 .index = { .avl_tree = { .root = NULL, .compar = appconfig_section_compare },
+                                            .rwlock = AVL_LOCK_INITIALIZER } };
+
+/* Note: this is not a CMocka group_test_setup/teardown pair. This is performed per-test.
+*/
+static struct web_client *setup_fresh_web_client()
+{
+    struct web_client *w = (struct web_client *)malloc(sizeof(struct web_client));
+    memset(w, 0, sizeof(struct web_client));
+    w->response.data = buffer_create(NETDATA_WEB_RESPONSE_INITIAL_SIZE);
+    w->response.header = buffer_create(NETDATA_WEB_RESPONSE_HEADER_SIZE);
+    w->response.header_output = buffer_create(NETDATA_WEB_RESPONSE_HEADER_SIZE);
+    strcpy(w->origin, "*"); // Simulate web_client_create_on_fd()
+    w->cookie1[0] = 0;      // Simulate web_client_create_on_fd()
+    w->cookie2[0] = 0;      // Simulate web_client_create_on_fd()
+    w->acl = 0x1f;          // Everything on
+    return w;
+}
+
+static void destroy_web_client(struct web_client *w)
+{
+    buffer_free(w->response.data);
+    buffer_free(w->response.header);
+    buffer_free(w->response.header_output);
+    free(w);
+}
+
+//////////////////////////// Test cases ///////////////////////////////////////////////////////////////////////////////
+
+static void only_root(void **state)
+{
+    (void)state;
+
+    if (localhost != NULL)
+        free(localhost);
+    localhost = malloc(sizeof(RRDHOST));
+
+    struct web_client *w = setup_fresh_web_client();
+    buffer_strcat(w->response.data, "GET / HTTP/1.1\r\n\r\n");
+
+    char debug[4096];
+    repr(debug, sizeof(debug), w->response.data->buffer, w->response.data->len);
+    printf("-> \"%s\"\n", debug);
+
+    //char expected_url_repr[4096];
+    //repr(expected_url_repr, sizeof(expected_url_repr), def->url_out_repr, strlen(def->url_out_repr));
+
+    expect_string(__wrap_mysendfile, filename, "/");
```

Comment:
An excessive check. Leads to "__wrap_mysendfile: '%s' parameter still has values that haven't been checked."

Reply:
I think the check is correct - the code being tested is calling the wrong result. It should treat everything outside of the `/api/v1` subtree as a file-request, so in this case it should call `mysendfile` with the argument supplied.